subprojects/model-core/src/main/java/org/gradle/api/internal/provider/ValueSupplier.java

```diff
@@ -210,35 +211,67 @@ public void visitProducerTasks(Action<? super Task> visitor) {
          * Creates a new side effect that ignores its argument
          * and instead always executes against the given {@code value}.
          */
-        static <T, A> SideEffect<T> fixed(A value, SideEffect<A> sideEffect) {
-            return new FixedSideEffect<>(value, sideEffect);
+        @Nullable
+        static <T, A> SideEffect<T> fixed(A value, @Nullable SideEffect<A> sideEffect) {
+            return FixedSideEffect.of(value, sideEffect);
+        }
+
+        /**
+         * Extracts the side effect from the {@code value} if any, and fixes it on that value.
+         * <p>
+         * If the value is {@link Value#isMissing() missing} then null is returned.
+         *
+         * @see #fixed
+         */
+        @Nullable
+        static <T, A> SideEffect<T> fixedFrom(Value<A> value) {
+            if (value.isMissing()) {
+                return null;
+            }
+
+            return fixed(value.getWithoutSideEffect(), value.getSideEffect());
+        }
+
+        /**
+         * Extracts the side effect from the {@code value} if any, and fixes it on that value.
+         * <p>
+         * If the execution time value is {@link ExecutionTimeValue#hasFixedValue() not fixed} then null is returned.
+         *
+         * @see #fixed
+         */
+        @Nullable
+        static <T, A> SideEffect<T> fixedFrom(ExecutionTimeValue<A> value) {
+            if (value.hasFixedValue()) {
```

Comment:
Shouldn't this be `!value.hasFixedValue()`? Is there a test missing?

Reply:
Nice catch! I will take a look

Reply:
Fixed!

Indeed, the tests could have been even more extensive. I updated the tests so that they catch this change.